resource/state_mover.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package resource
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+)
+
+// Implementation handler for a state move operation. After determining the
+// source resource is supported, this should encapsulate all data transformation
+// logic from a source resource to the current schema version of this
+// [Resource]. The [Resource] is connected to these implementations by
+// implementing the [ResourceWithMoveState] interface.
+//
+// This functionality is only available in Terraform 1.8 or later. It is invoked
+// when a configuration contains a `moved` configuration block where the source
+// resource type in the `from` argument differs from the target resource type in
+// the `to` argument, causing Terraform to call this provider operation and the
+// framework to route the request to this [Resource] as the target.
+//
+// Each implementation is responsible for determining whether the request should
+// be handled or skipped. The implementation is considered skipped by the
+// framework when the response contains no error diagnostics or state.
+// Otherwise, any error diagnostics or state data, will cause the framework to
+// consider the request completed and not call other [StateMover]
+// implementations. The framework will return an implementation not found error
+// if all implementations return responses without error diagnostics or state.
+type StateMover struct {
+	// SourceSchema is an optional schema for the intended source resource state
+	// and schema version. While not required, setting this will populate
+	// [MoveStateRequest.SourceState] when possible similar to other [Resource]
+	// types. State conversion errors based on this schema are only logged at
+	// DEBUG level as there may be multiple [StateMover] implementations on the
+	// same target resource for differing source resources.
+	//
+	// If not set, source state data is only available in
+	// [MoveStateRequest.SourceRawState].
+	SourceSchema *schema.Schema
```

Comment:
Should we mention explicitly that state conversion errors will not cause the `StateMover` function to be skipped? It's kind of implied but was something that came to my mind when reading the interface.